### PR TITLE
Add schema.Provider.TestReset to reset StopContext between tests

### PR DESCRIPTION
**Pull request by jbardin**
In normal operation, a provider is only created once and only ever canceled once, but in testing a ResourceProvider may be initialized and reused throughout multiple tests. 

This poses a problem when `Stop()` has been called on a provider and the StopContext has been used by the provider. 

This adds a `TestReset` method to `schema.Provider` to reset its internal state, currently only consisting of the internal Context. `Provider.TestReset` will then call Provider.MetaReset if it's not nil, which can be used to reset any state stored in the meta interface value which is especially important if it stored a StopContext value. 

The TestReset method will be automatically called after PreCheck in resource.Test.

### CODE REVIEW

**Comment by jbardin** [comment]
This is still a little awkward for provider, going to push an updated version...

**Comment by stack72** [comment]
Hi @jbardin 

I am still getting the errors using this branch...

```
% make testacc TEST=./builtin/providers/azurerm TESTARGS='-run=TestAccAzureRMStorageAccount_'
==> Checking that code complies with gofmt requirements...
go generate $(go list ./... | grep -v /terraform/vendor/)
2017/03/08 14:41:26 Generated command/internal_plugin_list.go
TF_ACC=1 go test ./builtin/providers/azurerm -v -run=TestAccAzureRMStorageAccount_ -timeout 120m
=== RUN   TestAccAzureRMStorageAccount_importBasic
--- FAIL: TestAccAzureRMStorageAccount_importBasic (68.12s)
	testing.go:267: Step 0 error: Error applying: 1 error(s) occurred:

		* azurerm_storage_account.testsa: 1 error(s) occurred:

		* azurerm_storage_account.testsa: Error creating Azure Storage Account 'unlikely23exst2accttnew': storage.AccountsClient#Create: Failure sending request: StatusCode=0 -- Original Error: Put https://management.azure.com/subscriptions/c0a607b2-6372-4ef3-abdb-dbe52a7b56ba/resourceGroups/testAccAzureRMSA-5659809402542612052/providers/Microsoft.Storage/storageAccounts/unlikely23exst2accttnew?api-version=2016-01-01: net/http: request canceled
=== RUN   TestAccAzureRMStorageAccount_basic
--- FAIL: TestAccAzureRMStorageAccount_basic (75.23s)
	testing.go:267: Step 0 error: Error applying: 1 error(s) occurred:

		* azurerm_storage_account.testsa: 1 error(s) occurred:

		* azurerm_storage_account.testsa: Error creating Azure Storage Account 'unlikely23exst2acct9sn2': storage.AccountsClient#Create: Failure sending request: StatusCode=0 -- Original Error: Put https://management.azure.com/subscriptions/c0a607b2-6372-4ef3-abdb-dbe52a7b56ba/resourceGroups/testAccAzureRMSA-5846784653319619894/providers/Microsoft.Storage/storageAccounts/unlikely23exst2acct9sn2?api-version=2016-01-01: net/http: request canceled
````

**Comment by jbardin** [comment]
@mitchellh: that last commit is the important one, and effects all uses of StopContext, not just in tests. Azure had problems simply because it was the first to make use of the new StopContext.

File: helper/resource/testing.go [hunk 1]
```diff
@@ -22,6 +22,12 @@ import (
 
 const TestEnvVar = "TF_ACC"
 
+// TestProvider is implemented by scheme.Provider to allow resetting the
+// provider state between tests.
```
Comment (mitchellh) [inline helper/resource/testing.go]: Nitpick but I'd reword the comment to say that this can be implemented by any ResourceProvider to have custom reset functionality, and note when TestReset is called. You can also make a note that helper/schema implements this.

File: helper/resource/testing.go [hunk 2]
```diff
@@ -333,6 +335,43 @@ func Test(t TestT, c TestCase) {
...
+		}
+	}
+
+	// reset the providers if needed
+	for k, pf := range ctxProviders {
+		// we can ignore any errors here, if we don't have a provider to reset
+		// the error will be handled later
+		p, _ := pf()
+		if p, ok := p.(TestProvider); ok {
+			err := p.TestReset()
```
Comment (mitchellh) [inline helper/resource/testing.go]: Do we have to do this for all providers or can we just do it as part of the constructed factory above (line 355)?

File: terraform/resource_provider_mock.go [hunk 3]
```diff
@@ -56,6 +56,8 @@ type MockResourceProvider struct {
 	ReadDataDiffFn                 func(*InstanceInfo, *ResourceConfig) (*InstanceDiff, error)
 	ReadDataDiffReturn             *InstanceDiff
 	ReadDataDiffReturnError        error
+	TestResetCalled                bool
+	TestResetError                 error
```
Comment (mitchellh) [inline terraform/resource_provider_mock.go]: Seems messy to add this interface here since the core doesn't care about it all. Could we modify the test in helper/resource to add it to a custom provider to verify it is called?

File: helper/resource/testing.go (hunk 1 above)
Comment (jbardin) [inline helper/resource/testing.go]: ah yes, I often forget that schema.Provider technically isn't the only ResourceProvider ;)

File: helper/resource/testing.go (hunk 2 above)
Comment (jbardin) [inline helper/resource/testing.go]: we could, but we need to range over both `c.Providers` and `c.ProviderFactories` above, and here I do the check once, rather than repeat it in both loops.

File: terraform/resource_provider_mock.go (hunk 3 above)
Comment (jbardin) [inline terraform/resource_provider_mock.go]: Makes sense. Removed the fields from terraform, and wrapped the mock provider with a TestProvider implementation.

File: helper/resource/testing.go (hunk 2 above)
Comment (mitchellh) [inline helper/resource/testing.go]: Ah I see. Yeah it is probably fine anyways. :+1:
